tsdb/series_collection.go

```diff
@@ -0,0 +1,334 @@
+	if ulength < uint(len(s.SeriesKeys)) {
+		s.SeriesKeys = s.SeriesKeys[ulength:]
+	}
+	if ulength < uint(len(s.Names)) {
+		s.Names = s.Names[ulength:]
+	}
+	if ulength < uint(len(s.Tags)) {
+		s.Tags = s.Tags[ulength:]
+	}
+	if ulength < uint(len(s.Types)) {
+		s.Types = s.Types[ulength:]
+	}
+	if ulength < uint(len(s.SeriesIDs)) {
+		s.SeriesIDs = s.SeriesIDs[ulength:]
+	}
+}
+
+// InvalidateAll causes all of the entries to become invalid.
+func (s *SeriesCollection) InvalidateAll(reason string) {
+	if s.Reason == "" {
+		s.Reason = reason
+	}
+	s.Dropped += uint64(len(s.Keys))
+	s.DroppedKeys = append(s.DroppedKeys, s.Keys...)
+	s.Truncate(0)
+}
+
+// ApplyConcurrentDrops will remove all of the dropped values during concurrent iteration. It should
+// not be called concurrently with any calls to Invalid.
+func (s *SeriesCollection) ApplyConcurrentDrops() {
+	state := s.getState(false)
+	if state == nil {
+		return
+	}
+
+	length, j := s.Length(), 0
+	for i := 0; i < length; i++ {
+		if _, ok := state.index[i]; ok {
+			s.Dropped++
+
+			if i < len(s.Keys) {
+				s.DroppedKeys = append(s.DroppedKeys, s.Keys[i])
+			}
+
+			continue
+		}
+
+		s.Copy(j, i)
+		j++
+	}
+	s.Truncate(j)
+
+	if s.Reason == "" {
+		s.Reason = state.reason
+	}
+
+	// clear concurrent state
+	atomic.StorePointer((*unsafe.Pointer)(unsafe.Pointer(&s.state)), nil)
+}
+
+// getState returns the SeriesCollection's concurrent state. If alloc is true and there
+// is no state, it will attempt to allocate one and set it. It is safe to call concurrently, but
+// not with ApplyConcurrentDrops.
+func (s *SeriesCollection) getState(alloc bool) *seriesCollectionState {
+	addr := (*unsafe.Pointer)(unsafe.Pointer(&s.state))
+
+	// fast path: load pointer and it already exists. always return the result if we can't alloc.
+	if ptr := atomic.LoadPointer(addr); ptr != nil || !alloc {
+		return (*seriesCollectionState)(ptr)
+	}
+
+	// nothing there. make a new state and try to swap it in.
+	atomic.CompareAndSwapPointer(addr, nil, unsafe.Pointer(new(seriesCollectionState)))
+
+	// reload the pointer. this way we always end up with the winner of the race.
+	return (*seriesCollectionState)(atomic.LoadPointer(addr))
+}
+
+// invalidIndex stages the index as invalid with the reason. It will be removed when
+// ApplyConcurrentDrops is called.
+func (s *SeriesCollection) invalidIndex(index int, reason string) {
+	state := s.getState(true)
+
+	state.mu.Lock()
+	if state.index == nil {
+		state.index = make(map[int]struct{})
+	}
+	state.index[index] = struct{}{}
+	if state.reason == "" {
+		state.reason = reason
+	}
+	state.mu.Unlock()
+}
+
+// PartialWriteError returns a PartialWriteError if any entries have been marked as invalid. It
+// returns an error to avoid `return collection.PartialWriteError()` always being non-nil.
+func (s *SeriesCollection) PartialWriteError() error {
+	if s.Dropped == 0 {
+		return nil
+	}
+	droppedKeys := bytesutil.SortDedup(s.DroppedKeys)
+	return PartialWriteError{
```

Comment:
I'm glad to see this is consistently a value vs mixed pointer / value. I verified Enterprise does not expect a pointer anywhere either 👍